proxy/enfuse/socket.go

```diff
@@ -0,0 +1,114 @@
+package enfuse
+
+import (
+	"bytes"
+	"github.com/enfabrica/enkit/lib/logger"
+	"github.com/gorilla/websocket"
+	"io"
+	"log"
+	"net"
+)
+
+var _ io.ReadWriter = &SocketShim{}
+
+// SocketShim is a simple wrapper that implements io.ReadWriter that writes and reads the full buffer while translating
+// payloads. If read from, it will strip the Prefix from the payload if it is present. If it is written to, it will
+// automatically append the Prefix.
+// in the future this could be nicely re-connectable with a buffer window.
+type SocketShim struct {
+	WebConn   *WebsocketLocker
+	Prefix    []byte
+	PrefixLen int
+}
+
+func (s SocketShim) Read(p []byte) (n int, err error) {
+	_, data, err := s.WebConn.ReadMessage()
+	if err != nil {
+		return 0, err
+	}
+	realData := bytes.Trim(data, "\x00")
+	copy(p, realData[s.PrefixLen:])
+	return len(realData[s.PrefixLen:]), io.EOF
+}
+
+func (s SocketShim) Write(p []byte) (n int, err error) {
+	// websockets always write the full buffer
+	realPayload := append(s.Prefix, p...)
+	return len(p), s.WebConn.WriteMessage(websocket.BinaryMessage, realPayload)
+}
+
+func NewSocketShim(strategy PayloadAppendStrategy, Conn *websocket.Conn) (*SocketShim, error) {
+	pLen, f := strategy()
+	pre, err := f()
```

Comment:
(future PR) this still smells to me - I suspect that PayloadAppendStrategy should actually be an interface

Reply:
Maybe, depending on how complex it gets